### PR TITLE
Proposal for delegating user permissions through API keys

### PR DESCRIPTION
**Pull request by nblumhardt**
[Rendered](https://github.com/nblumhardt/seq-tickets/blob/apikey-delegated-permissions/rfcs/0667-apikey-delegated-permissions.md)

### CODE REVIEW

File: rfcs/0667-apikey-delegated-permissions.md [hunk 1]
```diff
@@ -0,0 +1,124 @@
+# API key delegated permissions
+
+Instead of maintaining separate principal types and access rules for regular users vs API keys, switch to the model whereby a user creates API keys and delegates one or more of the user's permissions to it.
```
Comment (KodrAus) [inline rfcs/0667-apikey-delegated-permissions.md]: This reads a little clunky. What about:

_Unify access rules between regular users and API keys by switching to a model where API keys are delegated a subset of their owning user's permissions._

File: rfcs/0667-apikey-delegated-permissions.md [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+### Usage
+
+The summaries in this section highlight how the major use cases for API keys will be achieved in the new version.
+
+#### API keys for application log ingestion
+
+It's anticipated that API keys for application log ingestion will continue to be created and maintained by a system administrator, as shared (non-personal) API keys. This will prevent interruption of application logging when users come and go.
+
+A user may still create a personal API key and use it to ingest log events if desired - this essentially provides "self-service" API key creation. Using shared keys for ingesion is encouraged by recommendation only.
```
Comment (KodrAus) [inline rfcs/0667-apikey-delegated-permissions.md]: What do we mean by _encouraged by recommendation_ here?

File: rfcs/0667-apikey-delegated-permissions.md [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+Most reporting and integration tasks should be achievable using an API key with only `Read` and/or `Ingest` permissions - many integration scenarios are possible without requiring a `Write` API key.
+
+#### API keys for scripted system setup and administration
+
+API keys can be used for scripted setup in essentially the same manner as for reporting and integration, but with the `Write` or `Setup` permissions delegated to them.
+
+#### Controlling ingestion rate per API key
+
+When an administrative user browses _Settings > API keys_, by default only shared keys will be shown. By selectin the new _Show personal API keys_ checkbox, the details of API keys owned by all users will be included, showing the ingestion rate for personal as well as shared keys.
```
Comment (KodrAus) [inline rfcs/0667-apikey-delegated-permissions.md]: `s/selectin/selecting`

File: rfcs/0667-apikey-delegated-permissions.md (hunk 1 above)
Comment (nblumhardt) [reply rfcs/0667-apikey-delegated-permissions.md]: Thanks, I'll make the change :+1:

File: rfcs/0667-apikey-delegated-permissions.md (hunk 2 above)
Comment (nblumhardt) [reply rfcs/0667-apikey-delegated-permissions.md]: Will clarify - I meant to say that we'll recommend it but not take any steps to force it (it's a bit of a discretionary point).

**Pull request merged**